src/Generation/ServiceDetails.php

```diff
@@ -139,22 +142,41 @@ public function __construct(
         // Resource-names
         // Wildcard patterns are ignored.
         // A resource-name which has just a single wild-card pattern is ignored.
-        $messageResourceDefs = $this->methods
-            ->map(fn($x) => ProtoHelpers::getCustomOption($x->inputMsg, CustomOptions::GOOGLE_API_RESOURCEDEFINITION, ResourceDescriptor::class))
-            ->filter(fn($x) => !is_null($x));
-        $resourceRefs = $this->methods
-            ->flatMap(fn($x) => $x->allFields)
-            ->map(fn($x) => ProtoHelpers::getCustomOption($x->desc, CustomOptions::GOOGLE_API_RESOURCEREFERENCE, ResourceReference::class))
-            ->filter(fn($x) => !is_null($x));
-        $typeRefResourceDefs = $resourceRefs
-            ->filter(fn($x) => $x->getType() !== '' && $x->getType() !== '*')
-            ->map(fn($x) => $catalog->resourcesByType[$x->getType()]);
-        $childTypeRefResourceDefs = $resourceRefs
-            ->filter(fn($x) => $x->getChildType() !== '')
-            ->flatMap(fn($x) => $catalog->parentResourceByChildType->get($x->getChildType(), Vector::new([])));
-        $resourceDefs = $messageResourceDefs
-            ->concat($typeRefResourceDefs)
-            ->concat($childTypeRefResourceDefs)
+        $msgsSeen = Set::new();
+        $gatherMsgResDefs = null;
+        $gatherMsgResDefs = function(DescriptorProto $msg, int $level) use(&$gatherMsgResDefs, &$msgsSeen, $catalog): Vector {
+            if ($msgsSeen[$msg->desc->getFullname()]) {
+                return Vector::new([]);
+            }
+            $msgsSeen = $msgsSeen->add($msg->desc->getFullname());
+            // Only top-level resource-defs are included; matches monolith behaviour.
```

Comment:
Add a TODO(vNext) or TODO(monoilith) here and below? (so that future readers can better identify this like the other notes in this codebase)

Reply:
done